src/main/java/core/basesyntax/ArrayList.java

```diff
@@ -1,48 +1,156 @@
+    private Object[] values;
+    private int size;
+
+    public ArrayList() {
+        values = new Object[DEFAULT_ARRAY_CAPACITY];
+        size = 0;
+    }
+
     @Override
     public void add(T value) {
-
+        if (size == values.length) {
+            grow();
+        }
+        values[size] = value;
+        size = size + 1;
     }
 
     @Override
     public void add(T value, int index) {
+        if (index < 0) {
+            throw new ArrayListIndexOutOfBoundsException("Index can`t be lass than 0");
+        }
+        if (index > size) {
+            throw new ArrayListIndexOutOfBoundsException("Index "
+                    + index + " is bigger or equal the size ("
+                    + size + ")");
+        }
 
+        if (size == values.length) {
+            grow();
+        }
+
+        Object[] tempValues = new Object[values.length];
+        System.arraycopy(values, 0, tempValues, 0, index);
+        System.arraycopy(values, index, tempValues, index + 1, size - index);
+        tempValues[index] = value;
+        values = tempValues;
+        size = size + 1;
     }
 
     @Override
     public void addAll(List<T> list) {
+        if (list == null) {
+            throw new NullPointerException("Input can`t be null");
+        }
+
+        if (list.size() > values.length) {
+            grow(size + list.size());
+        }
 
+        for (int i = 0; i < list.size(); i++) {
+            add(list.get(i));
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        if (index < 0) {
+            throw new ArrayListIndexOutOfBoundsException("Index can`t be lass than 0");
+        }
+
+        if (index >= size) {
+            throw new ArrayListIndexOutOfBoundsException(
+                    "Index " + index
+                    + "is bigger or equal than size("
+                    + size + ")"
+            );
+        }
+        return (T) values[index];
     }
 
     @Override
     public void set(T value, int index) {
+        if (index < 0) {
+            throw new ArrayListIndexOutOfBoundsException("Index can`t be lass than 0");
+        }
 
+        if (index >= size) {
+            throw new ArrayListIndexOutOfBoundsException(
+                    "Index " + index
+                            + "is bigger or equal than size("
+                            + size + ")"
+            );
+        }
+        values[index] = value;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        if (index < 0) {
+            throw new ArrayListIndexOutOfBoundsException("Index can`t be lass than 0");
+        }
+
+        if (index >= size) {
+            throw new ArrayListIndexOutOfBoundsException(
+                    "Index " + index
+                            + "is bigger or equal than size("
+                            + size + ")"
+            );
+        }
+
+        T removedValue = (T) values[index];
```

Comment:
The logic of method could be simplified

checkIfIndexFits(index);
T removedValue = (T) values[index];
System.arraycopy(values, index + 1, values, index, size - index - 1);
size--;
return removedValue;